### PR TITLE
feat: json support issue #80

### LINKED ISSUE #80: Add json support
**Is your feature request related to a problem? Please describe.**
Proji currently only supports `TOML` as config file format. Since it's quite possible that proji will work with APIs in the future (#63), you should be able to export/import configs in `JSON` format as well.

**Describe the solution you'd like**
Add a flag to package related commands. `TOML` might stay the default format for configs but you could export configs to `JSON` with a `--to-json` flag or similar.

This is just an example solution. Other solutions are of course open for discussion!

### PR DESCRIPTION
**Pull request by FelixMohr**
### Summary

Hi @nikoksr, this pull request implements Json as a format for package configurations.

### Proposed Changes

-   Project import from a config file is able to import Json files. This will be done if the file extension is *.json*.
-   Project export now also supports Json format. This can be done with option *--json*.
-   Importing Json packages from a remote is also possible, as the modified method ``ImportPackageFromConfig`` is called by ``ImportPackageFromRemote``.
-   Unit tests to prove correctness of the implementation.

```
» proji package export --help                                                                                                                                                                                                      felixmohr@Felixs-MBP
Export one or more packages

Usage:
  proji package export LABEL [LABEL...] [flags]

Aliases:
  export, e

Flags:
  -a, --all                  Export all packages
  -d, --destination string   Destination for the export (default ".")
  -h, --help                 help for export
  -j, --json                 Export to a Json file
  -t, --template             Export a package config template
```
### Resolved Issues

Fixes #80 

Best regards,
Felix

### CODE REVIEW

File: cmd/packageExport.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ type packageExportCommand struct {
 }
 
 func newPackageExportCommand() *packageExportCommand {
-	var exportAll, template bool
+	var exportAll, template, json bool
```
Comment (nikoksr) [inline cmd/packageExport.go:20]: Maybe change variable name to `asJson`?

File: cmd/packageExport.go [hunk 2]
```diff
@@ -75,6 +75,7 @@ func newPackageExportCommand() *packageExportCommand {
 
 	cmd.Flags().BoolVarP(&template, "template", "t", false, "Export a package config template")
```
Comment (nikoksr) [inline cmd/packageExport.go:76]: Thinking about renaming the `template` flag to `example` or something.. not sure I'm happy with that name. Haven't used and looked at this piece of code for quite some time and it took me couple of minutes what was meant with `template`. 

What's your opinion on this?

File: pkg/domain/package.go [hunk 3]
```diff
@@ -47,5 +47,5 @@ type PackageService interface {
 	ImportPackageFromRemote(url *url.URL) (*Package, error)
 	ImportPackagesFromCollection(url *url.URL, exclude *regexp.Regexp) ([]*Package, error)
 
-	ExportPackageToConfig(pkg Package, destination string) (string, error)
+	ExportPackageToConfig(pkg Package, destination string, json bool) (string, error)
```
Comment (nikoksr) [inline pkg/domain/package.go:50]: Same here, `asJson` would be more expressive.

File: pkg/package/service/config.go [hunk 4]
```diff
@@ -39,29 +38,73 @@ func (ps packageService) ImportPackageFromConfig(path string) (*domain.Package,
...
+	bytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(bytes, pkg)
+	return err
 }
 
-func isConfigPathValid(path string) error {
+func checkConfigPath(path string) (json bool, err error) {
```
Comment (nikoksr) [inline pkg/package/service/config.go:94]: Pretty much same here. The named return value should be changed to `isJson`.

File: pkg/package/service/config.go [hunk 5]
```diff
@@ -39,29 +38,73 @@ func (ps packageService) ImportPackageFromConfig(path string) (*domain.Package,
...
+	isJson := strings.HasSuffix(path, ".json")
```
Comment (nikoksr) [inline pkg/package/service/config.go:95]: In that case this short variable declaration would turn into an assignment.

**Review by nikoksr — COMMENTED** [review]
Thanks a ton for your contribution! Looking good, just a couple of minor changes which make the code more declarative.

File: cmd/packageExport.go (hunk 2 above)
Comment (FelixMohr) [reply cmd/packageExport.go]: I agree this way it'll be easier to understand… I renamed the flag and the generated file.

File: cmd/packageExport.go [hunk 6]
```diff
@@ -73,18 +73,18 @@ func newPackageExportCommand() *packageExportCommand {
 		},
 	}
 
-	cmd.Flags().BoolVarP(&template, "template", "t", false, "Export a package config template")
+	cmd.Flags().BoolVarP(&template, "example", "e", false, "Export a package config example")
```
Comment (nikoksr) [inline cmd/packageExport.go:76]: Varibale name `template` should be changed to `example` too. Thanks for your work!

**Review by nikoksr — COMMENTED** [review]
Thanks for applying the changes!

File: cmd/packageExport.go (hunk 6 above)
Comment (FelixMohr) [reply cmd/packageExport.go]: Sorry I missed some stuff there :/ Should look better now

File: pkg/package/service/config.go [hunk 7]
```diff
@@ -39,29 +38,73 @@ func (ps packageService) ImportPackageFromConfig(path string) (*domain.Package,
...
+	}
+	return confName, ps.getJsonEncoderFunction(conf)(pkg)
+}
+
+func (ps packageService) getTomlEncoderFunction(conf *os.File) func(v interface{}) error {
+	return toml.NewEncoder(conf).Order(toml.OrderPreserve).Encode
+}
+
+func (ps packageService) getJsonEncoderFunction(conf *os.File) func(v interface{}) error {
+	return json.NewEncoder(conf).Encode
```
Comment (nikoksr) [inline pkg/package/service/config.go:64]: One last thing mate. Just noticed while testing your changes that `json` exports are not formatted. The file is just one line of json with everything in it. The config should be human-readable too. So, we need to add indentation to the json encoder like this: 

	encoder := json.NewEncoder(conf)
	encoder.SetIndent("", "  ")
	return encoder.Encode

**Review by nikoksr — COMMENTED** [review]
Last minor change, thanks so much for your efforts.

File: cmd/packageExport.go (hunk 6 above)
Comment (nikoksr) [reply cmd/packageExport.go]: No worries, great job!

File: pkg/package/service/config.go (hunk 7 above)
Comment (FelixMohr) [reply pkg/package/service/config.go]: sure thing

**Review by nikoksr — APPROVED** [review]
Great job. Your contribution is highly appreciated.

**Comment by nikoksr** [comment]
Will merge as soon as [v0.21.0](https://github.com/nikoksr/proji/milestone/25) is done. Thanks for your work :)

**Review by nikoksr — APPROVED** [review]

**Comment by georgettica** [comment]
I don''t want to step on any toes... but can I take ownership of this?
I can create a new PR with @FelixMohr changes and fix the merge conflicts

**Pull request closed without merging**